MarianInterface.cpp

```diff
@@ -28,58 +30,57 @@ marian::Ptr<marian::Options> MakeOptions(QString path_to_model_dir, translateLoc
 
 MarianInterface::MarianInterface(QString path_to_model_dir, translateLocally::marianSettings& settings, QObject *parent)
     : QObject(parent)
-    , service_(new marian::bergamot::Service(MakeOptions(path_to_model_dir, settings)))
-    , serial_(0)
-    , finished_(0)
-    , mymodel(path_to_model_dir) {}
+    , pendingInput_()
+    , pendingInputCount_(0)
+    , mymodel(path_to_model_dir) {
 
-void MarianInterface::translate(QString in) {
-    // Wait on future until Response is complete. Since the future doesn't have a callback or anything
-    // we should put all the processing in a background thread. Normally, if we have a future, we expect
-    // that future to have a method that allows to attach a callback, but this is reserved for c++20? c++22
-    // We have to copy any member variables we use (I'm looking at you QString input, because QString is copy-on-write)
-    auto translateAndSignal = [&](std::string &&input, std::size_t serial) {
-        using marian::bergamot::Response;
-        QString translation;
-
-        // If translating empty string, just pass through immediately but do
-        // update the finished_ variable to cancel out all previous translations.
-        if (input.empty()) {
-            translation = "";
-        } else {
-            std::future<marian::bergamot::Response> responseFuture = service_->translate(std::move(input));
+    worker_ = std::thread([&](marian::Ptr<marian::Options> options) {
+        marian::bergamot::Service service(options);
+
+        while (true) {
```

Comment:
This looks like busy waiting?

Reply:
So, it's a semaphore acquiring. Is there a chance for this to busy-wait? When we have a large translation pending and the user tries to submit a new one?

Reply:
Ok, nvm, github is showing the diff in a weird way.

So basically, you bypass the service queuing mechanism by creating your own worker and queue. Which is bad because we're reimplementing functionality from a broken upstream service we're using. But I understand